File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/video/FileOutputOptionsImpl.kt
package dev.hebei.camerax_android.video

import dev.hebei.camerax_android.CameraXImpl
import dev.hebei.camerax_android.PigeonApiFileOutputOptionsApi
import dev.hebei.camerax_android.common.api
import dev.hebei.camerax_android.common.fileImpl

class FileOutputOptionsImpl(impl: CameraXImpl) : PigeonApiFileOutputOptionsApi(impl) {
    override fun build(
        file: String, durationLimitMillis: Long?, fileSizeLimitBytes: Long?, location: android.location.Location?
    ): androidx.camera.video.FileOutputOptions {
        val builder = androidx.camera.video.FileOutputOptions.Builder(file.fileImpl).setLocation(location)
        if (durationLimitMillis != null) {
            builder.setDurationLimitMillis(durationLimitMillis)
        }
        if (fileSizeLimitBytes != null) {
            builder.setFileSizeLimit(fileSizeLimitBytes)
        }
        return builder.build()
    }

    override fun file(pigeon_instance: androidx.camera.video.FileOutputOptions): String {
        return pigeon_instance.file.api
    }
}

File: camerax_ios/lib/src/video/file_output_options_impl.dart
import 'dart:io';

import 'package:camerax_ios/src/api.dart';
import 'package:camerax_ios/src/common.dart';
import 'package:camerax_ios/src/video.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class FileOutputOptions$BuilderImpl implements FileOutputOptions$Builder {
  final FileOutputOptionsBuilderProxyApi api;

  FileOutputOptions$BuilderImpl.internal(this.api);

  @override
  Future<FileOutputOptions$Builder> setDurationLimitMillis(
    int durationLimitMillis,
  ) => api.setDurationLimitMillis(durationLimitMillis).then((e) => e.impl);

  @override
  Future<FileOutputOptions$Builder> setFileSizeLimit(int fileSizeLimitBytes) =>
      api.setFileSizeLimit(fileSizeLimitBytes).then((e) => e.impl);

  @override
  Future<FileOutputOptions$Builder> setLocation(Location? location) =>
      api.setLocation(location?.api).then((e) => e.impl);

  @override
  Future<FileOutputOptions> build() => api.build().then((e) => e.impl);
}

final class FileOutputOptionsImpl extends OutputOptionsImpl
    implements FileOutputOptions {
  @override
  final FileOutputOptionsProxyApi api;

  FileOutputOptionsImpl.internal(this.api);

  @override
  Future<File> getFile() => api.getFile().then((e) => e.fileImpl);
}

final class FileOutputOptionsChannelImpl extends FileOutputOptionsChannel {
  @override
  FileOutputOptions$Builder createBuilder(File file) {
    final api = FileOutputOptionsBuilderProxyApi(file: file.api);
    return FileOutputOptions$BuilderImpl.internal(api);
  }
}

extension FileOutputOptionsBuilderProxyApiX
    on FileOutputOptionsBuilderProxyApi {
  FileOutputOptions$Builder get impl =>
      FileOutputOptions$BuilderImpl.internal(this);
}

extension FileOutputOptionsX on FileOutputOptions {
  FileOutputOptionsProxyApi get api {
    final impl = this;
    if (impl is! FileOutputOptionsImpl) throw TypeError();
    return impl.api;
  }
}

extension FileOutputOptionsProxyApiX on FileOutputOptionsProxyApi {
  FileOutputOptions get impl => FileOutputOptionsImpl.internal(this);
}
